superset/views/core.py

```diff
@@ -2710,8 +2710,10 @@ def sql_json(self):
                     rendered_query,
                     return_results=True,
                     user_name=g.user.username if g.user else None)
+
+
             payload = json.dumps(
-                data,
+                apply_display_max_row_limit(data),
```

Comment:
In the case of a sync request, we may want to apply `DISPLAY_MAX_ROW` straight onto the SQL itself.

The reason we have 2 limits in the first place is because we want to store the 1M rows in the results backend (say S3) to allow for CSV download of the 1M row as a web request that reads from the s3 bucket, yet not crash the web browser.

Here's another related idea that's been discussed before around this to get data results faster to the user. It's the idea of doing a first dump after a `cursor.fetch_many(10000)`, we'd probably dump that so a specific cache key, and introduce a new query state: `SAMPLES_DELIVERED` (or something like that) and then fetch the rest `cursor.fetch_all()` before reaching `SUCCESS`, and dump the whole result set into another cache key. I think ultimately that's where we want to go.

Reply:
If we were applying `DISPLAY_MAX_ROW` directly onto the query, how would we know to show the user that we limited displayed results on the server side? We wouldn't know if 10k results were shown because there's only 10k results in the db or because the query was actually limited.

Reply:
I'm not sure if that's good idea but what about limiting to 10k+1 ?

Reply:
Hmm, so CSV download always makes an async request? If so then I'm happy to apply `DISPLAY_MAX_ROW` to the SQL, then make a follow up issue to track the `samples_delivered` enhancement

Reply:
CSV button is only shown after the query was executed, and if it did run async (has a results_key), it will download it from results cache.
https://github.com/apache/incubator-superset/blob/5470d10155e28c96b54615a36464c6c4afc0a38b/superset/views/core.py#L2754

If not async, it reruns the query, but takes a total shortcut which seems like it could be wrong to me in some cases, but not related to this PR... https://github.com/apache/incubator-superset/blob/5470d10155e28c96b54615a36464c6c4afc0a38b/superset/views/core.py#L2770

Reply:
After digging into this further, I wasn't able to find a clean way to apply the query limit for the first call of the query, but then remove it when the query is called again to generate the CSV.

The logic between async vs sync queries and UI display vs CSV download seem like they could all be simplified in some way, but I'm not quite sure the best way to refactor them at this point. Would it be alright to get this fix merged in for now, then to revisit all these performance enhancements/refactors in a later issue + PR?

Reply:
SGTM!